File: 102-binary-tree-level-order-traversal/binary-tree-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Pair {
    TreeNode node;
    int level;

    public Pair(TreeNode node, int level) {
        this.node = node;
        this.level = level;
    }
}
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if (root == null) return ans;

        Queue<Pair> queue = new LinkedList<>();
        queue.add(new Pair(root, 0));

        while (!queue.isEmpty()) {
            Pair currentPair = queue.poll();
            TreeNode node = currentPair.node;
            int level = currentPair.level;

            // Ensure ans has enough space for the current level
            if (ans.size() <= level) {
                ans.add(new ArrayList<>());
            }

            // Add the current node's value to the corresponding level
            ans.get(level).add(node.val);

            // Enqueue children with the level incremented by 1
            if (node.left != null) {
                queue.add(new Pair(node.left, level + 1));
            }
            if (node.right != null) {
                queue.add(new Pair(node.right, level + 1));
            }
        }

        return ans; 
    }
}

File: 102-binary-tree-level-order-traversal/binary-tree-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {

        if(root  == NULL)return {};

        queue<TreeNode*>q;
        q.push(root);
        
         vector<vector<int>>ans; 
        while(!q.empty())
        {
            int size = q.size();
             vector<int>temp;
            while(size--)
            {
                auto node = q.front();
                q.pop();
                temp.push_back(node->val);

                if(node->left !=NULL)
                {
                    q.push(node->left);
                }
               if(node->right!=NULL)
               {
                q.push(node->right);
               }
            }

            ans.push_back(temp);
        }

        return ans;

    }
};